File: samples/dotnet/DotNet/DotNet/GoogleTests.cs
using System;
using OpenQA.Selenium;
using OpenQA.Selenium.Appium;
using OpenQA.Selenium.Appium.Android;
using OpenQA.Selenium.Remote;
using NUnit.Framework;
using System.Threading;

namespace DotNet
{
    [TestFixture()]
    public class GoogleTests
    {
        private static AndroidDriver<AppiumWebElement> driver;
        private static string USER_NAME = "tester01";
        private static string API_KEY = "13e36639-92e3-411b-a067-3457b5dea573";
        private static string HOST_NAME = "ec2-54-226-177-179.compute-1.amazonaws.com";
        private static string PORT = "3001";
        private static Uri URL = null;
        private static TimeSpan INIT_TIMEOUT_SEC = TimeSpan.FromSeconds(180);
        private static TimeSpan IMPLICIT_TIMEOUT_SEC = TimeSpan.FromSeconds(10);
        private static string GOOGLE_URL = "https://mail.google.com";
        private static string GOOGLE_URL_HOMEPAGE = "https://mail.google.com/mail/mu/mp/4/#tl/priority/%5Esmartlabel_personal";
        private static string TRASH_URL = "https://mail.google.com/mail/mu/mp/4/#tl/Trash";
        private static string EMAIL_ADDRESS1 = "krypton.portal@gmail.com";
        private static string EMAIL_ADDRESS2 = "krypton.portal2@gmail.com";
        private static string PASSWORD = "Admin@123456";
        private static string SUBJECT1 = "It is a subject";
        private static string BODY1 = "It is a body";
        private static string MSG_NOEMAIL = "You have no mail here.";
        private static string MSG_NOEMAIL_PRIMARY = "You have no mail.\r\nPlease enjoy your day!";

        [SetUp]
        public void BeforeAll()
        {
            DesiredCapabilities capabilities = new DesiredCapabilities();
            capabilities.SetCapability("browserName", "chrome");
            capabilities.SetCapability("deviceName", "Nexus 5");
            capabilities.SetCapability("platformName", "Android");
            capabilities.SetCapability("platformVersion", "5.1.1");
            capabilities.SetCapability(CapabilityType.AcceptSslCertificates, true);
            URL = new Uri("http://" + USER_NAME + ":" + API_KEY + "@" + HOST_NAME + ":" + PORT + "/wd/hub");
            //URL = new Uri("http://127.0.0.1:4723/wd/hub");
            driver = new AndroidDriver<AppiumWebElement>(URL, capabilities, INIT_TIMEOUT_SEC);
            driver.Manage().Timeouts().ImplicitlyWait(IMPLICIT_TIMEOUT_SEC);
            driver.Manage().Cookies.DeleteAllCookies();
            Thread.Sleep(5000);
            driver.Navigate().Refresh();
        }
        [TearDown]
        public void AfterAll()
        {
            driver.Quit();
            Thread.Sleep(3000); // Mare sure Appium cleanup appium session
        }
        [Test()]
        public void GoogleLogin_Invalid()
        {
            driver.Navigate().GoToUrl(GOOGLE_URL);
            Sleep(5000);
            driver.FindElementById("next").Click();
            Sleep(2000);
            string errorMsg = driver.FindElementById("errormsg_0_Email").Text;
            Assert.AreEqual(errorMsg, "Please enter your email.");
            driver.FindElementById("Email").SendKeys("invalid_email@where.about");
            driver.FindElementById("next").Click();
            Sleep(2000);
            string errorMsg2 = driver.FindElementById("errormsg_0_Email").Text;
            Assert.AreEqual(errorMsg2,
                    "Sorry, Google doesn't recognize that email. Create an account using that address?");
        }

        [Test()]
        public void GoogleLogin_Valid()
        {
            Login(EMAIL_ADDRESS1, PASSWORD);
            Sleep(5000);
            string getURL = driver.Url;
            Assert.AreEqual(true, getURL.Contains("https://mail.google.com/mail"));
        }

        [Test()]
        public void GoogleComposeEmail()
        {
            Login(EMAIL_ADDRESS1, PASSWORD);
            Sleep(5000);
            // should compose email successfully by Gmail1
            driver.FindElementByXPath("//div[@id='views']//div[@aria-label='Compose']").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[@id='cmcc_composeto']//input[@id='composeto']").SendKeys(EMAIL_ADDRESS2);
            driver.FindElementByXPath("//input[@id='cmcsubj']").SendKeys(SUBJECT1);
            driver.FindElementById("cmcbody").SendKeys(BODY1);
            Sleep(1000);
            driver.FindElementByXPath("//div[@id='views']//div[text()='Send']").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[@aria-label='Menu']").Click();
            Sleep(1000);
            driver.FindElementByXPath("//span[text()='Sent Mail']").Click();
            Sleep(1000);

            // verify the new email exists on Sent Email folder on Gmail1
            string getText = driver.FindElementByXPath("//div[@class='fm']").Text;
            Assert.AreEqual(true, getText.Contains(SUBJECT1));
            Assert.AreEqual(true, getText.Contains(BODY1));
            driver.FindElementByXPath("//div[contains(@class,'m')]//div[@role='listitem'][1]").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[contains(@class,'V j hj') and text()='Details']").Click();
            Sleep(1000);
            string getSubject = driver.FindElementByXPath("//span[@class='kj']/span").Text;
            string getBody = driver.FindElementByXPath("//div[@class='Hi']").Text;
            string getSentEmail = driver.FindElementByXPath("//div[@class='Kg']/span").Text;
            string getToEmail = driver.FindElementByXPath("//div[@class='Kh']//span[@class='Hh']//span[@class='Kg ph']")
                    .Text;

            Assert.AreEqual(SUBJECT1, getSubject);
            Assert.AreEqual(BODY1, getBody);
            Assert.AreEqual(EMAIL_ADDRESS1, getSentEmail);
            Assert.AreEqual(EMAIL_ADDRESS2, getToEmail);

            // delete email on Sent Email and Trash folder on Gmail1
            driver.FindElementByXPath("//div[@id='cv__cntbt']//div[@class='V j Y Mm Kg']").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[@class='V j cb Ol'][1]").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[@class='M j T b hc  Vm Im']//div[@class='V j Xd']").Click();
            Sleep(2000);
            string getNoEmailMsg = driver.FindElementByXPath("//div[@class='Wl']").Text;

            Assert.AreEqual(MSG_NOEMAIL, getNoEmailMsg);

            driver.Navigate().GoToUrl(TRASH_URL);
            Sleep(5000);
            driver.FindElementByXPath("//div[@class='V j cb Ol'][1]").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[@class='M j T b hc q m']//div[text()='Delete forever']").Click();
            getNoEmailMsg = driver.FindElementByXPath("//div[@class='Wl']").Text;

            Assert.AreEqual(MSG_NOEMAIL, getNoEmailMsg);

            driver.FindElementByXPath("//div[@aria-label='Menu']").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[contains(@class,'V Y Rx Kg')]").Click();
            Sleep(1000);
            driver.FindElementByXPath("//button[@id='signout']").Click();
            Sleep(5000);

            driver.Navigate().GoToUrl(GOOGLE_URL);
            driver.Manage().Cookies.DeleteAllCookies();
            Thread.Sleep(3000);
            driver.Navigate().Refresh();


            // verify Gmail2 received a new email from Gmail1
            Login(EMAIL_ADDRESS2, PASSWORD);
            driver.Navigate().GoToUrl(GOOGLE_URL_HOMEPAGE);
            driver.Navigate().Refresh();
            Sleep(5000);

            getText = driver.FindElementByXPath("//div[@class='fm']").Text;
            Assert.AreEqual(true, getText.Contains(SUBJECT1));
            Assert.AreEqual(true, getText.Contains(BODY1));
            driver.FindElementByXPath("//div[contains(@class,'fm')]//div[@role='listitem'][1]").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[contains(@class,'V j hj') and text()='Details']").Click();
            Sleep(1000);
            getSubject = driver.FindElementByXPath("//span[@class='kj']/span").Text;
            getBody = driver.FindElementByXPath("//div[@class='Hi']").Text;
            getSentEmail = driver.FindElementByXPath("//div[@class='Kg']/span").Text;
            getToEmail = driver.FindElementByXPath("//div[@class='Kh']//span[@class='Hh']//span[@class='Kg ph']").Text;

            Assert.AreEqual(SUBJECT1, getSubject);
            Assert.AreEqual(BODY1, getBody);
            Assert.AreEqual(EMAIL_ADDRESS1, getSentEmail);
            Assert.AreEqual(EMAIL_ADDRESS2, getToEmail);

            // delete email on Primary Email and Trash folder on Gmail2
            driver.FindElementByXPath("//div[contains(@class,'M j T b hc Om o')]//div[text()='Primary'][1]").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[@class='V j cb Ol'][1]").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[@class='M j T b hc  Vm Im']//div[@class='V j Xd']").Click();
            getNoEmailMsg = driver.FindElementByXPath("//div[@class='Wl']").Text;

            Assert.AreEqual(MSG_NOEMAIL_PRIMARY, getNoEmailMsg);

            driver.Navigate().GoToUrl(TRASH_URL);
            Sleep(5000);
            driver.FindElementByXPath("//div[@class='V j cb Ol'][1]").Click();
            Sleep(1000);
            driver.FindElementByXPath("//div[@class='M j T b hc q m']//div[text()='Delete forever']").Click();
            getNoEmailMsg = driver.FindElementByXPath("//div[@class='Wl']").Text;

            Assert.AreEqual(MSG_NOEMAIL, getNoEmailMsg);
        }

        private void Login(string username, string password)
        {
            driver.Navigate().GoToUrl(GOOGLE_URL);
            Sleep(5000);
            driver.FindElementById("Email").SendKeys(username);
            driver.FindElementById("next").Click();
            Sleep(1000);
            driver.FindElementById("Passwd").SendKeys(password);
            driver.FindElementById("signIn").Click();
            Sleep(10000);
        }

        private void Sleep(double milliseconds)
        {
            TimeSpan tp = TimeSpan.FromMilliseconds(milliseconds);
            Thread.Sleep(tp);
        }
    }
}

File: samples/python/GoogleTests.py
import unittest
from unittest import TestCase
import time
from time import sleep
from appium import webdriver

class GoogleTests(TestCase):
    USER_NAME = "tester01"
    API_KEY = "13e36639-92e3-411b-a067-3457b5dea573"
    HOST_NAME = "test.kobiton.com"
    PORT = "3001"
    GOOGLE_URL = "https://mail.google.com"
    GOOGLE_URL_HOMEPAGE = "https://mail.google.com/mail/mu/mp/4/#tl/priority/%5Esmartlabel_personal"
    TRASH_URL = "https://mail.google.com/mail/mu/mp/4/#tl/Trash"
    EMAIL_ADDRESS1 = "krypton.portal@gmail.com"
    EMAIL_ADDRESS2 = "krypton.portal2@gmail.com"
    PASSWORD = "Admin@123456"
    SUBJECT1 = "It is a subject"
    BODY1 = "It is a body"
    MSG_NOEMAIL = "You have no mail here."
    MSG_NOEMAIL_PRIMARY = "You have no mail.\nPlease enjoy your day!"
    URL = ""
    SERVER_URL_LOCAL = "http://localhost:4723/wd/hub"
    desired_capabilities = {
        'browserName':'chrome',
        'platformName': 'Android',
        'platformVersion': '5.1.1',
        'deviceName': 'Nexus 5'
    }

    def setUp(self):
        self.URL = "http://" + self.USER_NAME + ":" + self.API_KEY + "@" + self.HOST_NAME + ":" + self.PORT + "/wd/hub"
        self.log ("WebDriver request initiated. Waiting for response, this may take a while.")
        self.driver = webdriver.Remote(command_executor=self.URL, desired_capabilities=self.desired_capabilities, browser_profile=None, proxy=None, keep_alive=False)
        self.driver.delete_all_cookies()
        self.driver.refresh()
    def tearDown(self):
        self.driver.quit()
        sleep(10)  # We need to wait appium server clean up environment

    def test_should_not_accept_empty_email_or_invalid_email(self):
        self.log ("should not accept empty and invalid email")
        self.driver.get(self.GOOGLE_URL)
        sleep(5)
        self.driver.find_element_by_id("next").click()
        sleep(2)
        errorMsg = self.driver.find_element_by_id("errormsg_0_Email").text
        self.assertEqual("Please enter your email.", errorMsg, "The error message '%s' displays." % errorMsg)
        self.driver.find_element_by_id("Email").send_keys("invalid_email@where.about")
        self.driver.find_element_by_id("next").click()
        sleep(2)
        errorMsg = self.driver.find_element_by_id("errormsg_0_Email").text
        self.assertEqual("Sorry, Google doesn't recognize that email. Create an account using that address?", errorMsg, "The error message '%s' displays." % errorMsg)
        pass
    def test_should_accept_valid_credentials(self):
        self.log ("should accept valid email credentials")
        self.login(self.EMAIL_ADDRESS1, self.PASSWORD)
        getURL = self.driver.current_url
        self.assertTrue("https://mail.google.com/mail" in getURL, "The current url contains text 'https://mail.google.com/mail'.")
        pass

    def test_should_compose_email_successfully(self):
        self.login(self.EMAIL_ADDRESS1, self.PASSWORD)
        self.log ("should compose email successfully by Gmail1")
        self.driver.find_element_by_xpath("//div[@id='views']//div[@aria-label='Compose']").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[@id='cmcc_composeto']//input[@id='composeto']").send_keys(self.EMAIL_ADDRESS2)
        self.driver.find_element_by_xpath("//input[@id='cmcsubj']").send_keys(self.SUBJECT1)
        self.driver.find_element_by_id("cmcbody").send_keys(self.BODY1)
        sleep(1)
        self.driver.find_element_by_xpath("//div[@id='views']//div[text()='Send']").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[@aria-label='Menu']").click()
        sleep(1)
        self.driver.find_element_by_xpath("//span[text()='Sent Mail']").click()
        sleep(1)

        self.log ("verify the new email exists on Sent Email folder on Gmail1")
        getText = self.driver.find_element_by_xpath("//div[@class='fm']").text

        self.assertTrue(self.SUBJECT1 in getText, "The subject '%s' exists." % self.SUBJECT1)
        self.assertTrue(self.BODY1 in getText, "The body '%s' exists." % self.BODY1)

        self.driver.find_element_by_xpath("//div[contains(@class,'m')]//div[@role='listitem'][1]").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[contains(@class,'V j hj') and text()='Details']").click()
        sleep(1)

        getSubject = self.driver.find_element_by_xpath("//span[@class='kj']/span").text
        getBody = self.driver.find_element_by_xpath("//div[@class='Hi']").text
        getSentEmail = self.driver.find_element_by_xpath("//div[@class='Kg']/span").text
        getToEmail = self.driver.find_element_by_xpath("//div[@class='Kh']//span[@class='Hh']//span[@class='Kg ph']").text

        self.assertEqual(self.SUBJECT1, getSubject, "The subject '%s' should be exist." % self.SUBJECT1)
        self.assertEqual(self.BODY1, getBody, "The body '%s' should be exist." % self.BODY1)
        self.assertEqual(self.EMAIL_ADDRESS1, getSentEmail, "The emailSend '%s' should be exist." % self.EMAIL_ADDRESS1)
        self.assertEqual(self.EMAIL_ADDRESS2, getToEmail, "The emailTo'%s' should be exist." % self.EMAIL_ADDRESS2)

        self.log ("delete email on Sent Email and Trash folder on Gmail1")
        self.driver.find_element_by_xpath("//div[@id='cv__cntbt']//div[@class='V j Y Mm Kg']").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[@class='V j cb Ol'][1]").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[@class='M j T b hc  Vm Im']//div[@class='V j Xd']").click()
        sleep(2)
        getNoEmailMsg = self.driver.find_element_by_xpath("//div[@class='Wl']").text

        self.assertEqual(self.MSG_NOEMAIL, getNoEmailMsg, "The message '%s' exists." % self.MSG_NOEMAIL)

        self.driver.get(self.TRASH_URL)
        sleep(5)
        self.driver.find_element_by_xpath("//div[@class='V j cb Ol'][1]").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[@class='M j T b hc q m']//div[text()='Delete forever']").click()
        getNoEmailMsg = self.driver.find_element_by_xpath("//div[@class='Wl']").text

        self.assertEqual(self.MSG_NOEMAIL, getNoEmailMsg, "The message '%s' exists." % self.MSG_NOEMAIL)

        self.driver.find_element_by_xpath("//div[@aria-label='Menu']").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[contains(@class,'V Y Rx Kg')]").click()
        sleep(1)
        self.driver.find_element_by_xpath("//button[@id='signout']").click()
        sleep(5)

        self.driver.get(self.GOOGLE_URL)
        self.driver.delete_all_cookies()
        self.driver.refresh()

        self.log ("verify Gmail2 received a new email from Gmail1")
        self.login(self.EMAIL_ADDRESS2, self.PASSWORD)
        self.driver.get(self.GOOGLE_URL_HOMEPAGE)
        self.driver.refresh()
        sleep(5)

        getText = self.driver.find_element_by_xpath("//div[@class='fm']").text
        self.assertTrue(self.SUBJECT1 in getText, "The subject '%s' exists." % self.SUBJECT1)
        self.assertTrue(self.BODY1 in getText, "The body '%s' exists." % self.BODY1)

        self.driver.find_element_by_xpath("//div[contains(@class,'fm')]//div[@role='listitem'][1]").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[contains(@class,'V j hj') and text()='Details']").click()
        sleep(1)

        getSubject = self.driver.find_element_by_xpath("//span[@class='kj']/span").text
        getBody = self.driver.find_element_by_xpath("//div[@class='Hi']").text
        getSentEmail = self.driver.find_element_by_xpath("//div[@class='Kg']/span").text
        getToEmail = self.driver.find_element_by_xpath("//div[@class='Kh']//span[@class='Hh']//span[@class='Kg ph']").text

        self.assertEqual(self.SUBJECT1, getSubject, "The subject '%s' should be exist." % self.SUBJECT1)
        self.assertEqual(self.BODY1, getBody, "The body '%s' should be exist." % self.BODY1)
        self.assertEqual(self.EMAIL_ADDRESS1, getSentEmail, "The emailSend '%s' should be exist." % self.EMAIL_ADDRESS1)
        self.assertEqual(self.EMAIL_ADDRESS2, getToEmail, "The emailTo'%s' should be exist." % self.EMAIL_ADDRESS2)

        self.log ("delete email on Primary Email and Trash folder on Gmail2")
        self.driver.find_element_by_xpath("//div[contains(@class,'M j T b hc Om o')]//div[text()='Primary'][1]").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[@class='V j cb Ol'][1]").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[@class='M j T b hc  Vm Im']//div[@class='V j Xd']").click()
        getNoEmailMsg = self.driver.find_element_by_xpath("//div[@class='Wl']").text

        self.assertEqual(self.MSG_NOEMAIL_PRIMARY, getNoEmailMsg, "The message '%s' exists." % self.MSG_NOEMAIL_PRIMARY)

        self.driver.get(self.TRASH_URL)
        sleep(5)
        self.driver.find_element_by_xpath("//div[@class='V j cb Ol'][1]").click()
        sleep(1)
        self.driver.find_element_by_xpath("//div[@class='M j T b hc q m']//div[text()='Delete forever']").click()
        getNoEmailMsg = self.driver.find_element_by_xpath("//div[@class='Wl']").text

        self.assertEqual(self.MSG_NOEMAIL, getNoEmailMsg, "The message '%s' exists." % self.MSG_NOEMAIL)
        pass

    def login(self, username=None, password=None):
        self.log ("login to Gmail with email: '%s' and password: '%s'" % (username, password))
        self.driver.get(self.GOOGLE_URL)
        sleep(5)
        self.driver.find_element_by_id("Email").send_keys(username)
        self.driver.find_element_by_id("next").click()
        sleep(3)
        self.driver.find_element_by_id("Passwd").send_keys(password)
        self.driver.find_element_by_id("signIn").click()
        sleep(10)

    def log(self, msg):
        print (time.strftime("%H:%M:%S") + ": " + msg)

if __name__ == "__main__":
    suite = unittest.TestSuite()
    suite.addTest(GoogleTests("test_should_not_accept_empty_email_or_invalid_email"))
    suite.addTest(GoogleTests("test_should_accept_valid_credentials"))
    suite.addTest(GoogleTests("test_should_compose_email_successfully"))
    unittest.TextTestRunner(verbosity=1).run(suite)
